### ISSUE TITLE
Implement Runge-Kutta Fehlberg Integrator in libRoadRunner

### ISSUE THREAD

**Issue by hsauro** [sys-bio/roadrunner#210]
We need a better Runge-Kutta integrator in addition to the standard RK4 we already have (See RK4Integrator.cpp and .h). Probably the easiest adaptive step size RK to implement is the RKF45 (Runge-Kutta Fehleberg). There are many algorithms for for described  online, in particular Algorithm 45 Page 579 of Numerical Analysis Asaithambi is quite clear (See attached page scan), also Algorithm 5.3 page 297 of Numerical Analysis Burden and Faires (We have it in the UW librarary but I also found a copy at http://ins.sjtu.edu.cn/people/mtang/textbook.pdf.

![image](https://cloud.githubusercontent.com/assets/1054990/10252443/5bd53274-68ec-11e5-9be5-2bf0ece80e49.png)

**Comment by 0u812** [comment]
Wait, why do we want an adaptive step size RK? Adaptive stepping is 
currently giving us a lot of grief.

**Comment by hsauro** [comment]
It's not the same approach as cvode does it, its a plain Runge Kutta that
simply ups or downs the step size depending on the error between a 4th and
5th order RK. I'm thinking it won't have the same issues. Of course the one
thing these can't do is event handling but that can wait for another time.

H

**Comment by hsauro** [comment]
Plus it should take only one day for Kiri to implement since the algorithm
is straight forward and he case base the code on the existing RK4
implementation.

H

**Comment by 0u812** [comment]
Okay, sounds good.

**Comment by 0u812** [comment]
@kirichoi I created a scaffold for you which includes the integrator class and registers it. All that's left for you to do is to tweak the `integrate` routine to match the fehlberg algorith. Here is the routine:
https://github.com/sys-bio/roadrunner/commit/78784cb195a0fcda6e5882fb451f52f53f6570c0#diff-a4ba0bab09c944e7602725530a5ea747R69

**Comment by 0u812** [comment]
Btw @kirichoi, you'll notice in the header I have a standard documentation style. Whenever you modify or create a class or method, please add/modify at least the following in the header file. Please use your initials for `@author`. Example
```
/**
     * @author KC
     * @brief A Runge-Kutta Fehlberg method for roadrunner
     * @details Kiri, please fill this in
     */
    class RK45Integrator: public Integrator {
    ...
};
```

Here is the header for your integrator:
https://github.com/sys-bio/roadrunner/blob/develop/source/RK45Integrator.h

**Comment by 0u812** [comment]
@kirichoi let me know if you have any questions. You should not need to modify the CMake or add any source files.

**Comment by 0u812** [comment]
@hsauro I removed your end of October label and replaced it with a milestone

**Comment by 0u812** [comment]
@kirichoi one more thing, you can test your code with test_integrator_api.exe in the bin directory. Let me know if you have questions regarding how to use it.

**Comment by 0u812** [comment]
@carolc24, are you familiar with the GitHub fork & pull model? The ideal way for you to make these changes is to fork roadrunner under your GitHub account, make your changes, and then submit a pull request to me.

If you have questions, I'm here Mon-Thurs this week, then I fly out to Salt Lake City for all of next week. You can always email me questions while I'm away.

**Comment by 0u812** [comment]
@carolc24 btw please make all changes on the develop branch. It's the default branch, so it should be selected automatically, but be aware that all changes need to be made on the develop branch. The master branch is for official releases only.

Here are Herbert's code contribution guidelines, you're pretty good about documenting your code so I'm sure you'll do fine, but here they are for reference:
https://docs.google.com/document/d/1xFyoP8ybHHdjPUKwK2c2emHdw-HUl8ohDwtaGJJIQKk/edit?pli=1

**Comment by carolc24** [comment]
Great, thanks.  I was having some internet problems this weekend, so I copied the scaffold code into a file on my computer instead of cloning the full repository.  I just made a new fork and I should be able to clone it onto my computer now.  If I have any questions I'll let you know.

**Comment by 0u812** [comment]
Awesome, glad to hear you're already working on it!

**Comment by carolc24** [comment]
Just committed changes, the current integrator is producing results that match cvode within about 1e-5.  The only problem is it stops early if it can't get to t_f in the number of steps inputted by the user.  It's easy to get around this by giving an arbitrarily large number of steps, in which case it will take exactly as many steps as it needs to go from t_0 to t_f.

**Comment by kirichoi** [comment]
I have inspected the code and fixed few issues. Commit 88d8164a7a44b5e7bdb21b7bf6244bd72b165420 contains the fix. Below is the performance of RKF45 compared to our CVODE.

![image](https://user-images.githubusercontent.com/7672320/29734530-8e111f40-89a7-11e7-9108-de21fe10d99c.png)
![image](https://user-images.githubusercontent.com/7672320/29734536-94a0c9c8-89a7-11e7-99b8-3f9d2666b6c6.png)

**Comment by 0u812** [comment]
Looks like this is causing [problems with the build](https://travis-ci.org/sys-bio/roadrunner/builds/268538397?utm_source=email&utm_medium=notification) (I think you have to include the header for numeric_limits, should be `#include <limits>`).

**Comment by kirichoi** [comment]
Should be fixed in d135a661e5233897256bacaf51ab980b3c5b5a6d. I don't know why VS cannot capture these kind of issues.

**Comment by 0u812** [comment]
Build is passing. Thanks!
